// recharts/lib/cartesian/Line.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Line = void 0;
exports.computeLinePoints = computeLinePoints;
var _react = _interopRequireWildcard(require("react"));
var _reactSmooth = _interopRequireDefault(require("react-smooth"));
var _isEqual = _interopRequireDefault(require("lodash/isEqual"));
var _clsx = _interopRequireDefault(require("clsx"));
var _Curve = require("../shape/Curve");
var _Dot = require("../shape/Dot");
var _Layer = require("../container/Layer");
var _LabelList = require("../component/LabelList");
var _ErrorBar = require("./ErrorBar");
var _DataUtils = require("../util/DataUtils");
var _ReactUtils = require("../util/ReactUtils");
var _Global = require("../util/Global");
var _ChartUtils = require("../util/ChartUtils");
var _ActivePoints = require("../component/ActivePoints");
var _SetTooltipEntrySettings = require("../state/SetTooltipEntrySettings");
var _CartesianGraphicalItemContext = require("../context/CartesianGraphicalItemContext");
var _GraphicalItemClipPath = require("./GraphicalItemClipPath");
var _chartLayoutContext = require("../context/chartLayoutContext");
var _PanoramaContext = require("../context/PanoramaContext");
var _lineSelectors = require("../state/selectors/lineSelectors");
var _hooks = require("../state/hooks");
var _SetLegendPayload = require("../state/SetLegendPayload");
var _excluded = ["type", "layout", "connectNulls"],
  _excluded2 = ["activeDot", "animateNewValues", "animationBegin", "animationDuration", "animationEasing", "connectNulls", "dot", "hide", "isAnimationActive", "label", "legendType", "xAxisId", "yAxisId"];
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var n = Object.getOwnPropertySymbols(e); for (r = 0; r < n.length; r++) o = n[r], -1 === t.indexOf(o) && {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (-1 !== e.indexOf(n)) continue; t[n] = r[n]; } return t; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); } // eslint-disable-next-line max-classes-per-file
/**
 * Internal props, combination of external props + defaultProps + private Recharts state
 */

/**
 * External props, intended for end users to fill in
 */

/**
 * Because of naming conflict, we are forced to ignore certain (valid) SVG attributes.
 */

var computeLegendPayloadFromAreaData = props => {
  var {
    dataKey,
    name,
    stroke,
    legendType,
    hide
  } = props;
  return [{
    inactive: hide,
    dataKey,
    type: legendType,
    color: stroke,
    value: name || dataKey,
    payload: props
  }];
};
function getTooltipEntrySettings(props) {
  var {
    dataKey,
    data,
    stroke,
    strokeWidth,
    fill,
    name,
    hide,
    unit
  } = props;
  return {
    dataDefinedOnItem: data,
    positions: undefined,
    settings: {
      stroke,
      strokeWidth,
      fill,
      dataKey,
      nameKey: undefined,
      name: (0, _ChartUtils.getTooltipNameProp)(name, dataKey),
      hide,
      type: props.tooltipType,
      color: props.stroke,
      unit
    }
  };
}
var generateSimpleStrokeDasharray = (totalLength, length) => {
  return "".concat(length, "px ").concat(totalLength - length, "px");
};
function repeat(lines, count) {
  var linesUnit = lines.length % 2 !== 0 ? [...lines, 0] : lines;
  var result = [];
  for (var i = 0; i < count; ++i) {
    result = [...result, ...linesUnit];
  }
  return result;
}
var getStrokeDasharray = (length, totalLength, lines) => {
  var lineLength = lines.reduce((pre, next) => pre + next);

  // if lineLength is 0 return the default when no strokeDasharray is provided
  if (!lineLength) {
    return generateSimpleStrokeDasharray(totalLength, length);
  }
  var count = Math.floor(length / lineLength);
  var remainLength = length % lineLength;
  var restLength = totalLength - length;
  var remainLines = [];
  for (var i = 0, sum = 0; i < lines.length; sum += lines[i], ++i) {
    if (sum + lines[i] > remainLength) {
      remainLines = [...lines.slice(0, i), remainLength - sum];
      break;
    }
  }
  var emptyLines = remainLines.length % 2 === 0 ? [0, restLength] : [restLength];
  return [...repeat(lines, count), ...remainLines, ...emptyLines].map(line => "".concat(line, "px")).join(', ');
};
function renderDotItem(option, props) {
  var dotItem;
  if (/*#__PURE__*/_react.default.isValidElement(option)) {
    dotItem = /*#__PURE__*/_react.default.cloneElement(option, props);
  } else if (typeof option === 'function') {
    dotItem = option(props);
  } else {
    var className = (0, _clsx.default)('recharts-line-dot', typeof option !== 'boolean' ? option.className : '');
    dotItem = /*#__PURE__*/_react.default.createElement(_Dot.Dot, _extends({}, props, {
      className: className
    }));
  }
  return dotItem;
}
var errorBarDataPointFormatter = (dataPoint, dataKey) => {
  return {
    x: dataPoint.x,
    y: dataPoint.y,
    value: dataPoint.value,
    // @ts-expect-error getValueByDataKey does not validate the output type
    errorVal: (0, _ChartUtils.getValueByDataKey)(dataPoint.payload, dataKey)
  };
};
class LineWithState extends _react.Component {
  constructor() {
    super(...arguments);
    _defineProperty(this, "state", {
      isAnimationFinished: true,
      totalLength: 0
    });
    _defineProperty(this, "id", (0, _DataUtils.uniqueId)('recharts-line-'));
    _defineProperty(this, "pathRef", node => {
      this.mainCurve = node;
    });
    _defineProperty(this, "handleAnimationEnd", () => {
      this.setState({
        isAnimationFinished: true
      });
      if (this.props.onAnimationEnd) {
        this.props.onAnimationEnd();
      }
    });
    _defineProperty(this, "handleAnimationStart", () => {
      this.setState({
        isAnimationFinished: false
      });
      if (this.props.onAnimationStart) {
        this.props.onAnimationStart();
      }
    });
  }
  componentDidMount() {
    if (!this.props.isAnimationActive) {
      return;
    }
    var totalLength = this.getTotalLength();
    this.setState({
      totalLength
    });
  }
  componentDidUpdate() {
    if (!this.props.isAnimationActive) {
      return;
    }
    var totalLength = this.getTotalLength();
    if (totalLength !== this.state.totalLength) {
      this.setState({
        totalLength
      });
    }
  }
  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.animationId !== prevState.prevAnimationId) {
      return {
        prevAnimationId: nextProps.animationId,
        curPoints: nextProps.points,
        prevPoints: prevState.curPoints
      };
    }
    if (nextProps.points !== prevState.curPoints) {
      return {
        curPoints: nextProps.points
      };
    }
    return null;
  }
  getTotalLength() {
    var curveDom = this.mainCurve;
    try {
      return curveDom && curveDom.getTotalLength && curveDom.getTotalLength() || 0;
    } catch (_unused) {
      return 0;
    }
  }
  renderDots(needClip, clipDot, clipPathId) {
    var {
      isAnimationActive
    } = this.props;
    if (isAnimationActive && !this.state.isAnimationFinished) {
      return null;
    }
    var {
      dot,
      points,
      dataKey
    } = this.props;
    var lineProps = (0, _ReactUtils.filterProps)(this.props, false);
    var customDotProps = (0, _ReactUtils.filterProps)(dot, true);
    var dots = points.map((entry, i) => {
      var dotProps = _objectSpread(_objectSpread(_objectSpread({
        key: "dot-".concat(i),
        r: 3
      }, lineProps), customDotProps), {}, {
        value: entry.value,
        dataKey,
        cx: entry.x,
        cy: entry.y,
        index: i,
        payload: entry.payload
      });
      return renderDotItem(dot, dotProps);
    });
    var dotsProps = {
      clipPath: needClip ? "url(#clipPath-".concat(clipDot ? '' : 'dots-').concat(clipPathId, ")") : null
    };
    return /*#__PURE__*/_react.default.createElement(_Layer.Layer, _extends({
      className: "recharts-line-dots",
      key: "dots"
    }, dotsProps), dots);
  }
  renderCurveStatically(points, needClip, clipPathId, props) {
    var _this$props = this.props,
      {
        type,
        layout,
        connectNulls
      } = _this$props,
      others = _objectWithoutProperties(_this$props, _excluded);
    var curveProps = _objectSpread(_objectSpread(_objectSpread({}, (0, _ReactUtils.filterProps)(others, true)), {}, {
      fill: 'none',
      className: 'recharts-line-curve',
      clipPath: needClip ? "url(#clipPath-".concat(clipPathId, ")") : null,
      points
    }, props), {}, {
      type,
      layout,
      connectNulls
    });
    return /*#__PURE__*/_react.default.createElement(_Curve.Curve, _extends({}, curveProps, {
      pathRef: this.pathRef
    }));
  }
  renderCurveWithAnimation(needClip, clipPathId) {
    var {
      points,
      strokeDasharray,
      isAnimationActive,
      animationBegin,
      animationDuration,
      animationEasing,
      animationId,
      animateNewValues,
      width,
      height
    } = this.props;
    var {
      prevPoints,
      totalLength
    } = this.state;
    return /*#__PURE__*/_react.default.createElement(_reactSmooth.default, {
      begin: animationBegin,
      duration: animationDuration,
      isActive: isAnimationActive,
      easing: animationEasing,
      from: {
        t: 0
      },
      to: {
        t: 1
      },
      key: "line-".concat(animationId),
      onAnimationEnd: this.handleAnimationEnd,
      onAnimationStart: this.handleAnimationStart
    }, _ref => {
      var {
        t
      } = _ref;
      if (prevPoints) {
        var prevPointsDiffFactor = prevPoints.length / points.length;
        var stepData = points.map((entry, index) => {
          var prevPointIndex = Math.floor(index * prevPointsDiffFactor);
          if (prevPoints[prevPointIndex]) {
            var prev = prevPoints[prevPointIndex];
            var interpolatorX = (0, _DataUtils.interpolateNumber)(prev.x, entry.x);
            var interpolatorY = (0, _DataUtils.interpolateNumber)(prev.y, entry.y);
            return _objectSpread(_objectSpread({}, entry), {}, {
              x: interpolatorX(t),
              y: interpolatorY(t)
            });
          }

          // magic number of faking previous x and y location
          if (animateNewValues) {
            var _interpolatorX = (0, _DataUtils.interpolateNumber)(width * 2, entry.x);
            var _interpolatorY = (0, _DataUtils.interpolateNumber)(height / 2, entry.y);
            return _objectSpread(_objectSpread({}, entry), {}, {
              x: _interpolatorX(t),
              y: _interpolatorY(t)
            });
          }
          return _objectSpread(_objectSpread({}, entry), {}, {
            x: entry.x,
            y: entry.y
          });
        });
        return this.renderCurveStatically(stepData, needClip, clipPathId);
      }
      var interpolator = (0, _DataUtils.interpolateNumber)(0, totalLength);
      var curLength = interpolator(t);
      var currentStrokeDasharray;
      if (strokeDasharray) {
        var lines = "".concat(strokeDasharray).split(/[,\s]+/gim).map(num => parseFloat(num));
        currentStrokeDasharray = getStrokeDasharray(curLength, totalLength, lines);
      } else {
        currentStrokeDasharray = generateSimpleStrokeDasharray(totalLength, curLength);
      }
      return this.renderCurveStatically(points, needClip, clipPathId, {
        strokeDasharray: currentStrokeDasharray
      });
    });
  }
  renderCurve(needClip, clipPathId) {
    var {
      points,
      isAnimationActive
    } = this.props;
    var {
      prevPoints,
      totalLength
    } = this.state;
    if (isAnimationActive && points && points.length && (!prevPoints && totalLength > 0 || !(0, _isEqual.default)(prevPoints, points))) {
      return this.renderCurveWithAnimation(needClip, clipPathId);
    }
    return this.renderCurveStatically(points, needClip, clipPathId);
  }
  render() {
    var _filterProps;
    var {
      hide,
      dot,
      points,
      className,
      xAxisId,
      yAxisId,
      top,
      left,
      width,
      height,
      isAnimationActive,
      id,
      needClip,
      layout
    } = this.props;
    if (hide || !points || !points.length) {
      return null;
    }
    var {
      isAnimationFinished
    } = this.state;
    var hasSinglePoint = points.length === 1;
    var layerClass = (0, _clsx.default)('recharts-line', className);
    var clipPathId = (0, _DataUtils.isNullish)(id) ? this.id : id;
    var {
      r = 3,
      strokeWidth = 2
    } = (_filterProps = (0, _ReactUtils.filterProps)(dot, false)) !== null && _filterProps !== void 0 ? _filterProps : {
      r: 3,
      strokeWidth: 2
    };
    var {
      clipDot = true
    } = (0, _ReactUtils.hasClipDot)(dot) ? dot : {};
    var dotSize = r * 2 + strokeWidth;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_Layer.Layer, {
      className: layerClass
    }, needClip && /*#__PURE__*/_react.default.createElement("defs", null, /*#__PURE__*/_react.default.createElement(_GraphicalItemClipPath.GraphicalItemClipPath, {
      clipPathId: clipPathId,
      xAxisId: xAxisId,
      yAxisId: yAxisId
    }), !clipDot && /*#__PURE__*/_react.default.createElement("clipPath", {
      id: "clipPath-dots-".concat(clipPathId)
    }, /*#__PURE__*/_react.default.createElement("rect", {
      x: left - dotSize / 2,
      y: top - dotSize / 2,
      width: width + dotSize,
      height: height + dotSize
    }))), !hasSinglePoint && this.renderCurve(needClip, clipPathId), /*#__PURE__*/_react.default.createElement(_ErrorBar.SetErrorBarPreferredDirection, {
      direction: layout === 'horizontal' ? 'y' : 'x'
    }, /*#__PURE__*/_react.default.createElement(_CartesianGraphicalItemContext.SetErrorBarContext, {
      xAxisId: xAxisId,
      yAxisId: yAxisId,
      data: points,
      dataPointFormatter: errorBarDataPointFormatter,
      errorBarOffset: 0
    }, this.props.children)), (hasSinglePoint || dot) && this.renderDots(needClip, clipDot, clipPathId), (!isAnimationActive || isAnimationFinished) && _LabelList.LabelList.renderCallByParent(this.props, points)), /*#__PURE__*/_react.default.createElement(_ActivePoints.ActivePoints, {
      activeDot: this.props.activeDot,
      points: points,
      mainColor: this.props.stroke,
      itemDataKey: this.props.dataKey
    }));
  }
}
var defaultLineProps = {
  activeDot: true,
  animateNewValues: true,
  animationBegin: 0,
  animationDuration: 1500,
  animationEasing: 'ease',
  connectNulls: false,
  dot: true,
  fill: '#fff',
  hide: false,
  isAnimationActive: !_Global.Global.isSsr,
  label: false,
  legendType: 'line',
  stroke: '#3182bd',
  strokeWidth: 1,
  xAxisId: 0,
  yAxisId: 0
};
function LineImpl(props) {
  var {
    needClip
  } = (0, _GraphicalItemClipPath.useNeedsClip)(props.xAxisId, props.yAxisId);
  var {
    height,
    width,
    left,
    top
  } = (0, _chartLayoutContext.useOffset)();
  var layout = (0, _chartLayoutContext.useChartLayout)();
  var isPanorama = (0, _PanoramaContext.useIsPanorama)();
  var points = (0, _hooks.useAppSelector)(state => (0, _lineSelectors.selectLinePoints)(state, props.xAxisId, props.yAxisId, isPanorama, {
    dataKey: props.dataKey,
    data: props.data
  }));
  var updateId = (0, _chartLayoutContext.useUpdateId)();
  if (layout !== 'horizontal' && layout !== 'vertical') {
    // Cannot render Line in an unsupported layout
    return null;
  }
  var {
      activeDot = defaultLineProps.activeDot,
      animateNewValues = defaultLineProps.animateNewValues,
      animationBegin = defaultLineProps.animationBegin,
      animationDuration = defaultLineProps.animationDuration,
      animationEasing = defaultLineProps.animationEasing,
      connectNulls = defaultLineProps.connectNulls,
      dot = defaultLineProps.dot,
      hide = defaultLineProps.hide,
      isAnimationActive = defaultLineProps.isAnimationActive,
      label = defaultLineProps.label,
      legendType = defaultLineProps.legendType,
      xAxisId = defaultLineProps.xAxisId,
      yAxisId = defaultLineProps.yAxisId
    } = props,
    everythingElse = _objectWithoutProperties(props, _excluded2);
  return /*#__PURE__*/_react.default.createElement(LineWithState, _extends({}, everythingElse, {
    connectNulls: connectNulls,
    dot: dot,
    activeDot: activeDot,
    animateNewValues: animateNewValues,
    animationBegin: animationBegin,
    animationDuration: animationDuration,
    animationEasing: animationEasing,
    isAnimationActive: isAnimationActive,
    hide: hide,
    label: label,
    legendType: legendType,
    xAxisId: xAxisId,
    yAxisId: yAxisId,
    animationId: updateId,
    points: points,
    layout: layout,
    height: height,
    width: width,
    left: left,
    top: top,
    needClip: needClip
  }));
}
function computeLinePoints(_ref2) {
  var {
    layout,
    xAxis,
    yAxis,
    xAxisTicks,
    yAxisTicks,
    dataKey,
    bandSize,
    displayedData
  } = _ref2;
  return displayedData.map((entry, index) => {
    // @ts-expect-error getValueByDataKey does not validate the output type
    var value = (0, _ChartUtils.getValueByDataKey)(entry, dataKey);
    if (layout === 'horizontal') {
      return {
        x: (0, _ChartUtils.getCateCoordinateOfLine)({
          axis: xAxis,
          ticks: xAxisTicks,
          bandSize,
          entry,
          index
        }),
        y: (0, _DataUtils.isNullish)(value) ? null : yAxis.scale(value),
        value,
        payload: entry
      };
    }
    return {
      x: (0, _DataUtils.isNullish)(value) ? null : xAxis.scale(value),
      y: (0, _ChartUtils.getCateCoordinateOfLine)({
        axis: yAxis,
        ticks: yAxisTicks,
        bandSize,
        entry,
        index
      }),
      value,
      payload: entry
    };
  });
}
class Line extends _react.PureComponent {
  render() {
    // Report all props to Redux store first, before calling any hooks, to avoid circular dependencies.
    return /*#__PURE__*/_react.default.createElement(_CartesianGraphicalItemContext.CartesianGraphicalItemContext, {
      type: "line",
      data: this.props.data,
      xAxisId: this.props.xAxisId,
      yAxisId: this.props.yAxisId,
      zAxisId: 0,
      dataKey: this.props.dataKey
      // line doesn't stack
      ,
      stackId: undefined,
      hide: this.props.hide,
      barSize: undefined
    }, /*#__PURE__*/_react.default.createElement(_SetLegendPayload.SetLegendPayload, {
      legendPayload: computeLegendPayloadFromAreaData(this.props)
    }), /*#__PURE__*/_react.default.createElement(_SetTooltipEntrySettings.SetTooltipEntrySettings, {
      fn: getTooltipEntrySettings,
      args: this.props
    }), /*#__PURE__*/_react.default.createElement(LineImpl, this.props));
  }
}
exports.Line = Line;
_defineProperty(Line, "displayName", 'Line');
_defineProperty(Line, "defaultProps", defaultLineProps);
